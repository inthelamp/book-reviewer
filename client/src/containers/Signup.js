import { useState } from 'react'
import { Form, Button } from 'react-bootstrap'
import { Redirect } from 'react-router-dom'
import axios from 'axios'
import Message from '../components/Message'
import './Signup.css'

const Signup = () => {
  const [email, setEmail] = useState('')
  const [userName, setUserName] = useState('') 
  const [password, setPassword] = useState('')
  const [confirmPassword, setConfirmPassword] = useState('')  
  const [message, setMessage] = useState('')
  const [messageStyle, setMessageStyle] = useState('')
  const [isSignedUp, setIsSignedUp] = useState(false)

  const isValid = () => {
    return email.length > 0 && userName.length > 0 && password.length > 0 && confirmPassword.length > 0 && password === confirmPassword
  }  

  const onSubmit = (e) => {
    e.preventDefault();

    axios
    .post(process.env.REACT_APP_SERVER_BASE_URL + '/users/signup', {email, userName, password, confirmPassword})
    .then(() => {
        setIsSignedUp(true)
    })
    .catch ((error) => {
        setMessageStyle('error_message')      
        if (error.response) {
          setMessage(error.response.data.message)                
        } else {
          console.log('Error', error.message)
        }
    })
  }

  const onChangePassword = (e) => {    
    setPassword(e.currentTarget.value)

    if (confirmPassword) {
      if (e.currentTarget.value !== confirmPassword) {
        let message = 'These passwords don’t match.'
        setMessage(message)
        setMessageStyle('error_message')
      } else {     
        setMessage('')
        setMessageStyle('')            
      }
    } else {
      setMessage('')
      setMessageStyle('')      
    }
  } 

  const onChangeConfirmPassword = (e) => {    
    setConfirmPassword(e.currentTarget.value)

    if (e.currentTarget.value !== password) {
      let message = 'These passwords don’t match.'
      setMessage(message)
      setMessageStyle('error_message')
    } else {     
      setMessage('')
      setMessageStyle('')            
    }
  }

  if (isSignedUp) {
    return <Redirect to={{
            pathname: '/activate',
            state: { email }
          }} />
  }

  return (
    <div className='signup'> 
      <Form onSubmit={onSubmit}>
      <Form.Group size='lg' controlId='email'>
        <Form.Label>Email</Form.Label>
        <Form.Control
          autoFocus
          type='email'
          value={email}
          onChange={(e) => setEmail(e.currentTarget.value)}
      />
      </Form.Group>      
      <Form.Group size='lg' controlId='userName'>
        <Form.Label>User Name</Form.Label>
        <Form.Control
          type='text'
          value={userName}
          onChange={(e) => setUserName(e.currentTarget.value)}
      />      
      </Form.Group>
      <Form.Group size='lg' controlId='password'>
        <Form.Label>Password</Form.Label>
        <Form.Control
          type='password'
          value={password}
          onChange={onChangePassword}
        />
        <Form.Label>Password Confirm</Form.Label>        
        <Form.Control
          type='password'
          value={confirmPassword}
          onChange={onChangeConfirmPassword}
        />          
      </Form.Group>
      <Button size='lg' type='submit' disabled={!isValid()} >
        Sign Up
      </Button>
      <Message message={message} messageStyle={messageStyle}/>
    </Form>      
   </div> 
  )
}

export default Signup
