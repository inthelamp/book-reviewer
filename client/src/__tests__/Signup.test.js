import { BrowserRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import Signup from '../containers/Signup'


describe('Component: Signup', () => {
    it('Component: Signup rendering correctly', () => {  

        const tree = renderer
            .create(<Router><Signup /></Router>)
            .toJSON();
        expect(tree).toMatchSnapshot()
    })
})