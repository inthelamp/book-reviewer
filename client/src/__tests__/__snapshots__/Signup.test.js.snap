// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Signup Component: Signup rendering correctly 1`] = `
<div
  className="signup"
>
  <form
    className=""
    onSubmit={[Function]}
  >
    <div
      class="mb-3"
      size="lg"
    >
      <label
        className="form-label"
        htmlFor="email"
      >
        Email
      </label>
      <input
        autoFocus={true}
        className="form-control"
        id="email"
        onChange={[Function]}
        type="email"
        value=""
      />
    </div>
    <div
      className="mb-3"
      size="lg"
    >
      <label
        className="form-label"
        htmlFor="userName"
      >
        User Name
      </label>
      <input
        className="form-control"
        id="userName"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="mb-3"
      size="lg"
    >
      <label
        className="form-label"
        htmlFor="password"
      >
        Password
      </label>
      <input
        className="mb-3 form-control"
        id="password"
        onChange={[Function]}
        type="password"
        value=""
      />
      <label
        className="form-label"
        htmlFor="password"
      >
        Password Confirm
      </label>
      <input
        className="form-control"
        id="password"
        onChange={[Function]}
        type="password"
        value=""
      />
    </div>
    <button
      className="btn btn-primary btn-lg"
      disabled={true}
      type="submit"
    >
      Sign Up
    </button>
    
  </form>
</div>
`;
